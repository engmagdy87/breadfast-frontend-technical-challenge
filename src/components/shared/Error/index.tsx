import React from "react";
import { Box, Button, Typography, useMediaQuery } from "@mui/material";
import Image from "next/image";
import theme from "utils/theme";
import FullScreenCenterContentWrapper from "components/shared/FullScreenCenterContentWrapper";

interface ErrorProps {
  retryAction: () => void;
}

const Error: React.FC<ErrorProps> = ({ retryAction }) => {
  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));

  return (
    <FullScreenCenterContentWrapper>
      <Image
        src="/warning.png"
        alt="Warning icon"
        width={100}
        height={100}
        style={{ height: "auto" }}
      />
      <Typography
        variant={isMobile ? "h5" : "h4"}
        color="primary"
        sx={{
          fontWeight: 600,
          mt: 3,
        }}
      >
        Something Went Wrong
      </Typography>
      <Typography
        variant={isMobile ? "h6" : "h5"}
        color="primary"
        sx={{
          mt: 3,
        }}
      >
        Please try again!
      </Typography>
      <Button
        variant="contained"
        color="primary"
        disableElevation
        size="large"
        sx={{ mt: 3, fontWeight: 700 }}
        onClick={retryAction}
      >
        Try again
      </Button>
    </FullScreenCenterContentWrapper>
  );
};

export default Error;
