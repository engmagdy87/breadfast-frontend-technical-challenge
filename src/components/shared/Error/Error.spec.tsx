import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Error from ".";

jest.mock("../../../components/shared/InformativeScreen", () => {
  return function MockInformativeScreen({
    icon,
    iconAlt,
    title,
    subtitle,
    buttonText,
    buttonTestId,
    buttonAction,
  }: {
    icon: string;
    iconAlt: string;
    title: string;
    subtitle: string;
    buttonText: string;
    buttonTestId: string;
    buttonAction: () => void;
  }) {
    return (
      <div data-testid="informative-screen">
        <img src={icon} alt={iconAlt} data-testid="error-icon" />
        <h1>{title}</h1>
        <p>{subtitle}</p>
        <button data-testid={buttonTestId} onClick={buttonAction}>
          {buttonText}
        </button>
      </div>
    );
  };
});

describe("Error Component", () => {
  const mockRetryAction = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders the InformativeScreen component with correct props", async () => {
    render(<Error retryAction={mockRetryAction} />);

    expect(screen.getByTestId("informative-screen")).toBeInTheDocument();

    const icon = screen.getByTestId("error-icon");
    expect(icon).toHaveAttribute("src", "/warning.png");
    expect(icon).toHaveAttribute("alt", "Warning icon");

    expect(screen.getByText("Something Went Wrong")).toBeInTheDocument();
    expect(screen.getByText("Please try again!")).toBeInTheDocument();

    const retryButton = screen.getByTestId("try-again-btn");
    expect(retryButton).toBeInTheDocument();
    expect(retryButton).toHaveTextContent("Try Again");
  });

  it("calls retryAction when Try Again button is clicked", () => {
    render(<Error retryAction={mockRetryAction} />);

    fireEvent.click(screen.getByTestId("try-again-btn"));

    expect(mockRetryAction).toHaveBeenCalledTimes(1);
  });

  it("passes accessibility checks", () => {
    const { container } = render(<Error retryAction={mockRetryAction} />);

    const icon = screen.getByTestId("error-icon");
    expect(icon).toHaveAttribute("alt");

    const button = screen.getByTestId("try-again-btn");
    expect(button).toBeEnabled();
  });

  it("matches snapshot", () => {
    const { container } = render(<Error retryAction={mockRetryAction} />);
    expect(container).toMatchSnapshot();
  });

  it("maintains consistent rendering when rerendered", () => {
    const { rerender } = render(<Error retryAction={mockRetryAction} />);
    const initialButton = screen.getByTestId("try-again-btn");

    rerender(<Error retryAction={mockRetryAction} />);
    const rerenderButton = screen.getByTestId("try-again-btn");

    expect(initialButton).toBe(rerenderButton);
  });
});

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeEnabled(): R;
    }
  }
}
