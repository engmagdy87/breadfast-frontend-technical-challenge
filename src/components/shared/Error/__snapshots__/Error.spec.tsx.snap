// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Component matches snapshot 1`] = `
<div>
  <div
    data-testid="informative-screen"
  >
    <img
      alt="Warning icon"
      data-testid="error-icon"
      src="/warning.png"
    />
    <h1>
      Something Went Wrong
    </h1>
    <p>
      Please try again!
    </p>
    <button
      data-testid="try-again-btn"
    >
      Try Again
    </button>
  </div>
</div>
`;
